### PR TITLE
fix(web): OSK resource-loading promise's font-wait ⛲

### PR DESCRIPTION
**Pull request by jahorton**
Enhances the `Promise` setup used to await OSK resource loading to also properly wait for a keyboard's fonts to load.

Rather than use techniques from KMW's past, I've instead utilized a (comparatively) newer Web feature:  [`FontFace`](https://developer.mozilla.org/en-US/docs/Web/API/FontFace).  It's newer than when the setup we _previously_ used was first written, but it's supported on all browser versions KMW currently supports and provides a _much_ cleaner way to achieve the same goals - it even directly provides a Promise for font-loading completion!

This changeset includes a tweak to ensure that there's always a layout-refresh once the keyboard's fonts load; what's small in one font can be large in another, after all.  

This admittedly corresponds better to the original design/logic behind similar handling in KMW 15.0 and before than what it's replacing from 16.0.

----

For comparison, in versions of KeymanWeb up to and including 15.0, when a keyboard's stylesheets were added, we'd wait for font loads like this:

https://github.com/keymanapp/keyman/blob/2dcc11448ee6e715fcfbbbaf6701a29f449e5817/web/source/osk/visualKeyboard.ts#L1365-L1368

That method's comments & signature:

https://github.com/keymanapp/keyman/blob/2dcc11448ee6e715fcfbbbaf6701a29f449e5817/web/source/osk/visualKeyboard.ts#L1725-L1731

We dropped this technique when we dropped the use of touch-alias elements.

@keymanapp-test-bot skip

See #10534; the best user test I could come up with is implemented therein.

### CODE REVIEW

**Comment by mcdurdin** [comment]
Given this changes significant core functionality in the font loading, I think we need user tests.

File: web/src/engine/dom-utils/src/stylesheets.ts [hunk 1]
```diff
@@ -132,9 +152,7 @@ export class StylesheetManager {
         if(this.doCacheBusting) {
           ttf = this.cacheBust(ttf);
         }
-        s=s+'src:url(\''+ttf+'\') format(\'truetype\');';
-      } else {
-        return null;
+        source = "url('"+ttf+"') format('truetype')";
       }
     } else {
       var s0 = [];
```
Comment (mcdurdin) [inline web/src/engine/dom-utils/src/stylesheets.ts:158]:
```suggestion
```

Is `s0` used any more?

File: web/src/engine/dom-utils/src/stylesheets.ts [hunk 2]
```diff
@@ -144,42 +162,55 @@ export class StylesheetManager {
         // with embedded ttf or woff.  svg mostly works so is a better initial
         // choice on the Android browser.
         if(svg != '') {
-          s0.push("url('"+svg+"') format('svg')");
+          source = "url('"+svg+"') format('svg')";
         }
```
Comment (mcdurdin) [inline web/src/engine/dom-utils/src/stylesheets.ts:166]: SVG is dead for fonts. We shouldn't be perpetuating if we are changing anyway.

l.161 references Android 4.2, 4.3 which we no longer support. We need to clean that up too.

And I see a reference to EOT also. Even deaderer

File: web/src/engine/dom-utils/src/stylesheets.ts [hunk 3]
```diff
@@ -144,42 +162,55 @@ export class StylesheetManager {
...
 
         if(ttf != '') {
-          s0.push("url('"+ttf+"') format('truetype')");
+          source = "url('"+ttf+"') format('truetype')";
         }
 
         if(svg != '') {
-          s0.push("url('"+svg+"') format('svg')");
+          source = "url('"+svg+"') format('svg')";
         }
```
Comment (mcdurdin) [inline web/src/engine/dom-utils/src/stylesheets.ts:186]: Ditto time for svg to go

File: web/src/engine/dom-utils/src/stylesheets.ts [hunk 4]
```diff
@@ -144,42 +162,55 @@ export class StylesheetManager {
         // with embedded ttf or woff.  svg mostly works so is a better initial
         // choice on the Android browser.
         if(svg != '') {
-          s0.push("url('"+svg+"') format('svg')");
+          source = "url('"+svg+"') format('svg')";
         }
 
         if(woff != '') {
-          s0.push("url('"+woff+"') format('woff')");
+          source = "url('"+woff+"') format('woff')";
```
Comment (mcdurdin) [inline web/src/engine/dom-utils/src/stylesheets.ts:169]: Ideally we should be supporting woff2 as well but that probably should be a separate PR

File: web/src/engine/dom-utils/src/stylesheets.ts [hunk 5]
```diff
@@ -132,9 +152,7 @@ export class StylesheetManager {
         if(this.doCacheBusting) {
           ttf = this.cacheBust(ttf);
         }
-        s=s+'src:url(\''+ttf+'\') format(\'truetype\');';
-      } else {
-        return null;
+        source = "url('"+ttf+"') format('truetype')";
```
Comment (mcdurdin) [inline web/src/engine/dom-utils/src/stylesheets.ts:155]: Since we are fixing this up should we be escaping the `ttf` string for safety?

File: web/src/engine/dom-utils/src/stylesheets.ts [hunk 6]
```diff
@@ -144,42 +162,55 @@ export class StylesheetManager {
...
+     *
+     * Note:  this could probably wholesale-replace the stylesheet!
+     * Would need: `document.fonts.add(fontFace)` - does not have to wait for the load() Promise.
+     *
+     * For now, we're using this solely to detect when the font has been succesfully loaded.
+     */
+    const fontFace = new FontFace(fd.family, source);
+    const loadPromise = fontFace.load();
+    this.fontPromises.push(loadPromise);
+    loadPromise.then(() => this.fontPromises = this.fontPromises.filter((entry) => entry != loadPromise));
```
Comment (mcdurdin) [inline web/src/engine/dom-utils/src/stylesheets.ts:212]: Wouldn't it be cleaner to just clear `fontPromises` after `allSettled()` on l.62 or so?

File: web/src/engine/dom-utils/src/stylesheets.ts [hunk 7]
```diff
@@ -197,15 +228,15 @@ export class StylesheetManager {
    *
    * @param   {string}  href   path to stylesheet file
    */
-  linkExternalSheet(href: string): void {
+  linkExternalSheet(href: string, force?: boolean): HTMLStyleElement {
```
Comment (mcdurdin) [inline web/src/engine/dom-utils/src/stylesheets.ts:231]: There's a change to this function signature but the change doesn't seem to be used?

File: web/src/engine/dom-utils/src/stylesheets.ts [hunk 8]
```diff
@@ -197,15 +228,15 @@ export class StylesheetManager {
...
     try {
-      if(document.querySelector("link[href="+JSON.stringify(href)+"]") != null) {
+      if(!force && document.querySelector("link[href="+JSON.stringify(href)+"]") != null) {
         // We've already linked this stylesheet, don't do it again
-        return;
+        return null;
       }
     } catch(e) {
       // We've built an invalid href, somehow?
```
Comment (mcdurdin) [inline web/src/engine/dom-utils/src/stylesheets.ts:238]: Suggests we should be logging

**Review by mcdurdin — CHANGES_REQUESTED** [review]

File: web/src/engine/dom-utils/src/stylesheets.ts (hunk 7 above)
Comment (jahorton) [reply web/src/engine/dom-utils/src/stylesheets.ts:231]: It comes into play in the followup - #10507.  Admittedly, the two were part of a single batch of changes at first, with me splitting them into separate chunks later.

If need be, I can undo it here and re-implement in that PR, which is a direct follow-up.

File: web/src/engine/dom-utils/src/stylesheets.ts (hunk 1 above)
Comment (jahorton) [reply web/src/engine/dom-utils/src/stylesheets.ts:158]: ... I thought I did remove it; good catch.

**Comment by jahorton** [comment]
> Given this changes significant core functionality in the font loading, I think we need user tests.

OK... so... how would I define user tests in a way that would be sufficiently clear?  That's the main reason I didn't spec any - setup instructions aren't hard, but setting the _expectations_ clearly, in an unambiguous way... that's where the trick lies.

File: web/src/engine/dom-utils/src/stylesheets.ts (hunk 7 above)
Comment (mcdurdin) [reply web/src/engine/dom-utils/src/stylesheets.ts:231]: no that's fine, I see the changes there.

**Comment by mcdurdin** [comment]
> > Given this changes significant core functionality in the font loading, I think we need user tests.
> 
> OK... so... how would I define user tests in a way that would be sufficiently clear? That's the main reason I didn't spec any - setup instructions aren't hard, but setting the _expectations_ clearly, in an unambiguous way... that's where the trick lies.

Agree this is tricky. Need to find an example (or craft one) where the font change is obvious I guess?

File: web/src/engine/dom-utils/src/stylesheets.ts (hunk 6 above)
Comment (jahorton) [reply web/src/engine/dom-utils/src/stylesheets.ts:212]: There's technically a chance for another stylesheet to come in at a later point that would need a `Promise` not currently included.  It's best not to assume that there will only be one call to the `allLoadedPromise()` method - largely because, in #10507, documentation-keyboards do end up needing a separate call to the method.  (Though... I might actually be able to restructure that a bit.)

To me, this is the cleanest way - each Promise ensures that it's only included in the `allLoadedPromise` when it's actually still loading, bowing out as soon as it's completed.  If new ones rotate in and old ones rotate out in the meantime, that all comes out in the wash with the current approach.  That is, so long as I add `.catch` handling too - _that_ part is certainly an important "catch" here.

Comment (mcdurdin) [reply web/src/engine/dom-utils/src/stylesheets.ts:212]: Secondary, do `Promise.all()` (and cousins) handle the scenario of the array changing while they are waiting?

**Comment by jahorton** [comment]
Requesting a review given that the related user test seen within #10534 has passed.

**Review by mcdurdin — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.257-alpha](https://keyman.com/downloads/releases/alpha/17.0.257)
